src/metadata/index.rs

```diff
@@ -0,0 +1,116 @@
+use elasticsearch::cat::CatIndicesParts;
+use elasticsearch::params::Time;
+use elasticsearch::{Elasticsearch, Error};
+use std::collections::HashMap;
+use tokio::sync::RwLock;
+
+use crate::{exporter_metrics::SUBSYSTEM_REQ_HISTOGRAM, Exporter};
+
+pub(crate) type IndexToMetadata = RwLock<IndexDataMap>;
+
+pub(crate) type IndexDataMap = HashMap<String, IndexMetadata>;
+
+/// Current Index metadata
+#[derive(Debug, Clone)]
+pub struct IndexMetadata {
+    pub last_hearbeat: chrono::DateTime<chrono::Utc>,
+}
+
+impl IndexMetadata {
+    /// Initialize Index metadata
+    pub fn new() -> Self {
+        Self {
+            last_hearbeat: chrono::Utc::now(),
+        }
+    }
+
+    /// Set heartbeat
+    pub fn reset_heartbeat(&mut self) -> &mut Self {
+        self.last_hearbeat = chrono::Utc::now();
+        self
+    }
+}
+
+/// Current Index metadata
+#[derive(Debug, Deserialize, Default)]
+pub(crate) struct CatIndexData {
+    /// index name
+    pub(crate) index: String,
+}
+
+pub(crate) async fn build(client: &Elasticsearch) -> Result<IndexDataMap, Error> {
+    info!("Elasticsearch: fetching indices metadata");
+    let index_data_map = client
+        .cat()
+        .indices(CatIndicesParts::None)
+        .format("json")
+        .h(&["index"])
+        // Return local information, do not retrieve the state from master node (default: false)
+        .local(true)
+        .time(Time::Ms)
+        .send()
+        .await?
+        .json::<Vec<CatIndexData>>()
+        .await?
+        .into_iter()
+        .map(|value| (value.index, IndexMetadata::new()))
+        .collect();
+
+    Ok(index_data_map)
+}
+
+#[inline]
+fn update_map(old: &mut IndexDataMap, new: IndexDataMap) {
+    for (key, new_metadata) in new {
+        let _ = old.entry(key).or_insert(new_metadata).reset_heartbeat();
+    }
+}
+
+#[allow(unused)]
+pub(crate) async fn poll(exporter: Exporter) {
+    let start = tokio::time::Instant::now();
+
+    let mut interval =
+        tokio::time::interval_at(start, exporter.options().exporter_metadata_refresh_interval);
+
+    loop {
+        let _ = interval.tick().await;
+
+        let timer = SUBSYSTEM_REQ_HISTOGRAM
+            .with_label_values(&["/_cat/indices", exporter.cluster_name()])
+            .start_timer();
+
+        match build(&exporter.client()).await {
+            Ok(new_metadata) => {
+                update_map(&mut *exporter.index_metadata().write().await, new_metadata)
+            }
+            Err(e) => {
+                error!("poll metadata metrics err {}", e);
+            }
+        }
+
+        timer.observe_duration();
+    }
+}
+
+#[test]
```

Comment:
I think you should put this into a dedicated `tests` module, no? Otherwise, it'll be part of the final binary.

Reply:
From docs https://doc.rust-lang.org/book/ch11-03-test-organization.html:
>The `#[cfg(test)]` annotation on the tests module tells Rust to compile and run the test code only when you run `cargo test`, not when you run `cargo build`.

At least that's how I understand it, I might be wrong though 🤷🏻

Reply:
I am still not sure :smile: anyhow moved under tests as it is cleaner anyway